File: typescript/src/utils/http.ts
/**
 * HTTP client utilities
 */

import axios, { AxiosInstance, AxiosRequestConfig, AxiosResponse, AxiosError } from 'axios';
import {
  NorthRelayError,
  AuthenticationError,
  ValidationError,
  QuotaExceededError,
  RateLimitError,
  NotFoundError,
  ServerError,
  NetworkError,
} from '../errors';
import type { ErrorResponse, RateLimitInfo } from '../types';
import { SDK_VERSION } from '../version';

export interface HttpClientConfig {
  baseURL: string;
  apiKey: string;
  timeout: number;
}

export class HttpClient {
  private client: AxiosInstance;
  private rateLimitInfo: RateLimitInfo | null = null;

  constructor(config: HttpClientConfig) {
    this.client = axios.create({
      baseURL: config.baseURL,
      timeout: config.timeout,
      headers: {
        'Authorization': `Bearer ${config.apiKey}`,
        'Content-Type': 'application/json',
        'User-Agent': `NorthRelay-SDK/${SDK_VERSION}`,
      },
    });

    // Response interceptor to extract rate limit headers
    this.client.interceptors.response.use(
      (response) => {
        this.extractRateLimitInfo(response);
        return response;
      },
      (error) => {
        if (error.response) {
          this.extractRateLimitInfo(error.response);
        }
        return Promise.reject(error);
      }
    );
  }

  public getRateLimitInfo(): RateLimitInfo | null {
    return this.rateLimitInfo;
  }

  public async request<T>(config: AxiosRequestConfig): Promise<T> {
    try {
      const response = await this.client.request<T>(config);
      return response.data;
    } catch (error) {
      throw this.handleError(error as AxiosError<ErrorResponse>);
    }
  }

  public async get<T>(url: string, config?: AxiosRequestConfig): Promise<T> {
    return this.request<T>({ ...config, method: 'GET', url });
  }

  public async post<T>(url: string, data?: unknown, config?: AxiosRequestConfig): Promise<T> {
    return this.request<T>({ ...config, method: 'POST', url, data });
  }

  public async put<T>(url: string, data?: unknown, config?: AxiosRequestConfig): Promise<T> {
    return this.request<T>({ ...config, method: 'PUT', url, data });
  }

  public async patch<T>(url: string, data?: unknown, config?: AxiosRequestConfig): Promise<T> {
    return this.request<T>({ ...config, method: 'PATCH', url, data });
  }

  public async delete<T>(url: string, config?: AxiosRequestConfig): Promise<T> {
    return this.request<T>({ ...config, method: 'DELETE', url });
  }

  private extractRateLimitInfo(response: AxiosResponse): void {
    const headers = response.headers;
    const limit = headers['x-ratelimit-limit'];
    const remaining = headers['x-ratelimit-remaining'];
    const reset = headers['x-ratelimit-reset'];

    if (limit && remaining && reset) {
      this.rateLimitInfo = {
        limit: parseInt(limit, 10),
        remaining: parseInt(remaining, 10),
        reset: parseInt(reset, 10),
      };
    }
  }

  private handleError(error: AxiosError<ErrorResponse>): Error {
    // Network error (no response)
    if (!error.response) {
      return new NetworkError(error.message || 'Network request failed');
    }

    const { status, data } = error.response;
    const errorData = data?.error;

    // API error with structured response
    if (errorData) {
      switch (errorData.code) {
        case 'AUTHENTICATION_FAILED':
        case 'INVALID_API_KEY':
        case 'MISSING_API_KEY':
          return new AuthenticationError(
            errorData.message,
            errorData.fix_action
          );

        case 'VALIDATION_ERROR':
          return new ValidationError(
            errorData.message,
            errorData.validationErrors
          );

        case 'QUOTA_EXCEEDED':
          return new QuotaExceededError(errorData.message);

        case 'RATE_LIMIT_EXCEEDED':
          const retryAfter = parseInt(error.response.headers['retry-after'] || '60', 10);
          return new RateLimitError(errorData.message, retryAfter);

        case 'NOT_FOUND':
        case 'TEMPLATE_NOT_FOUND':
        case 'DOMAIN_NOT_FOUND':
        case 'WEBHOOK_NOT_FOUND':
          return new NotFoundError(errorData.message);

        case 'INTERNAL_ERROR':
        case 'SMTP_ERROR':
        case 'SMTP_AUTH_ERROR':
        case 'SMTP_CONNECTION_FAILED':
          return new ServerError(errorData.message);

        default:
          return new NorthRelayError(
            errorData.message,
            errorData.code,
            status,
            errorData.fix_action,
            errorData.docs_url
          );
      }
    }

    // Generic HTTP error
    switch (status) {
      case 401:
        return new AuthenticationError('Authentication failed');
      case 404:
        return new NotFoundError('Resource');
      case 429:
        const retryAfter = parseInt(error.response.headers['retry-after'] || '60', 10);
        return new RateLimitError('Rate limit exceeded', retryAfter);
      case 500:
      case 502:
      case 503:
      case 504:
        return new ServerError('Server error occurred');
      default:
        return new NorthRelayError(
          error.message || 'An error occurred',
          'UNKNOWN_ERROR',
          status
        );
    }
  }
}

File: python/northrelay/utils/http.py
"""HTTP client with retry and rate limiting"""

import httpx
from typing import Any, Optional, TypeVar, cast
from datetime import datetime

from northrelay.exceptions import (
    AuthenticationError,
    ScopeError,
    ValidationError,
    QuotaExceededError,
    RateLimitError,
    NotFoundError,
    ServerError,
    NetworkError,
)
from northrelay.types import RateLimitInfo

T = TypeVar("T")


class HttpClient:
    """HTTP client for NorthRelay API with authentication and error handling"""

    def __init__(
        self,
        base_url: str,
        api_key: str,
        timeout: float = 30.0,
    ):
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.client = httpx.AsyncClient(
            base_url=self.base_url,
            timeout=timeout,
            headers={
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json",
                "User-Agent": "northrelay-python/1.1.0",
            },
        )
        self._rate_limit_info: Optional[RateLimitInfo] = None

    def get_rate_limit_info(self) -> Optional[RateLimitInfo]:
        """Get current rate limit information from last response"""
        return self._rate_limit_info

    def _update_rate_limit(self, response: httpx.Response) -> None:
        """Extract rate limit info from response headers"""
        headers = response.headers
        
        limit = headers.get("x-ratelimit-limit")
        remaining = headers.get("x-ratelimit-remaining")
        reset = headers.get("x-ratelimit-reset")

        if limit or remaining or reset:
            self._rate_limit_info = RateLimitInfo(
                limit=int(limit) if limit else None,
                remaining=int(remaining) if remaining else None,
                reset=datetime.fromtimestamp(int(reset)) if reset else None,
            )

    def _handle_error_response(self, response: httpx.Response) -> None:
        """Convert HTTP error responses to exceptions"""
        status_code = response.status_code
        
        try:
            error_data = response.json()
            error_message = error_data.get("message", error_data.get("error", response.text))
        except Exception:
            error_message = response.text or f"HTTP {status_code} error"

        # 401 - Authentication error
        if status_code == 401:
            raise AuthenticationError(error_message)

        # 403 - Scope / permission error
        if status_code == 403:
            raise ScopeError(error_message)

        # 400 - Validation error
        if status_code == 400:
            errors = error_data.get("errors") if isinstance(error_data, dict) else None
            raise ValidationError(error_message, errors=errors)

        # 404 - Not found
        if status_code == 404:
            raise NotFoundError(error_message)

        # 429 - Rate limit or quota
        if status_code == 429:
            retry_after = response.headers.get("retry-after")
            
            # Check if it's quota exceeded vs rate limit
            if "quota" in error_message.lower():
                raise QuotaExceededError(error_message)
            
            raise RateLimitError(
                error_message,
                retry_after=int(retry_after) if retry_after else None,
            )

        # 5xx - Server errors
        if 500 <= status_code < 600:
            raise ServerError(error_message, status_code=status_code)

        # Other errors
        response.raise_for_status()

    async def request(
        self,
        method: str,
        path: str,
        **kwargs: Any,
    ) -> httpx.Response:
        """Make HTTP request with error handling"""
        try:
            response = await self.client.request(method, path, **kwargs)
            self._update_rate_limit(response)
            
            if not response.is_success:
                self._handle_error_response(response)
            
            return response
            
        except httpx.RequestError as e:
            raise NetworkError(f"Network error: {e}", cause=e)

    async def get(self, path: str, **kwargs: Any) -> dict[str, Any]:
        """GET request"""
        response = await self.request("GET", path, **kwargs)
        return response.json()

    async def post(self, path: str, json: Any = None, **kwargs: Any) -> dict[str, Any]:
        """POST request"""
        response = await self.request("POST", path, json=json, **kwargs)
        return response.json()

    async def patch(self, path: str, json: Any = None, **kwargs: Any) -> dict[str, Any]:
        """PATCH request"""
        response = await self.request("PATCH", path, json=json, **kwargs)
        return response.json()

    async def put(self, path: str, json: Any = None, **kwargs: Any) -> dict[str, Any]:
        """PUT request"""
        response = await self.request("PUT", path, json=json, **kwargs)
        return response.json()

    async def delete(self, path: str, **kwargs: Any) -> dict[str, Any]:
        """DELETE request"""
        response = await self.request("DELETE", path, **kwargs)
        return response.json() if response.content else {}

    async def close(self) -> None:
        """Close HTTP client"""
        await self.client.aclose()

    async def __aenter__(self) -> "HttpClient":
        return self

    async def __aexit__(self, *args: Any) -> None:
        await self.close()
